pkg/rtc/uptrackmanager.go

```diff
@@ -0,0 +1,664 @@
+	// all_participants takes precedence
+	if permissions.AllParticipants {
+		// everything is allowed, nothing else to do
+		return
+	}
+
+	// if track_permissions is empty, nobody is allowed to subscribe to any tracks
+	if permissions.TrackPermissions != nil && len(permissions.TrackPermissions) == 0 {
+		for trackSid, _ := range u.publishedTracks {
+			u.subscriptionPermissions[trackSid] = []string{}
+		}
+
+		for _, ti := range u.pendingTracks {
+			u.subscriptionPermissions[ti.Sid] = []string{}
+		}
+		return
+	}
+
+	// per participant permissions
+	for _, trackPerms := range permissions.TrackPermissions {
+		// all_tracks get precedence
+		if trackPerms.AllTracks {
+			// no restriction for participant, nothing else to do
+			for trackSid, _ := range u.publishedTracks {
+				u.subscriptionPermissions[trackSid] = append(u.subscriptionPermissions[trackSid], trackPerms.ParticipantSid)
+			}
+
+			for _, ti := range u.pendingTracks {
+				u.subscriptionPermissions[ti.Sid] = append(u.subscriptionPermissions[ti.Sid], trackPerms.ParticipantSid)
+			}
+		} else {
+			for _, trackSid := range trackPerms.TrackSids {
+				u.subscriptionPermissions[trackSid] = append(u.subscriptionPermissions[trackSid], trackPerms.ParticipantSid)
+			}
+		}
+	}
+}
+
+func (u *UptrackManager) hasPermission(trackSid string, subscriberID string) bool {
+	allowed, ok := u.subscriptionPermissions[trackSid]
+	if !ok {
+		return true
+	}
+
+	for _, sid := range allowed {
+		if sid == subscriberID {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (u *UptrackManager) maybeAddPendingSubscription(trackSid string, sub types.Participant) {
+	subscriberID := sub.ID()
+
+	pending := u.pendingSubscriptions[trackSid]
+	for _, sid := range pending {
+		if sid == subscriberID {
+			// already pending
+			return
+		}
+	}
+
+	u.pendingSubscriptions[trackSid] = append(u.pendingSubscriptions[trackSid], subscriberID)
+	go sub.SubscriptionPermissionUpdate(u.params.SID, trackSid, false)
+}
+
+func (u *UptrackManager) maybeRemovePendingSubscription(trackSid string, sub types.Participant) {
+	subscriberID := sub.ID()
+
+	pending := u.pendingSubscriptions[trackSid]
+	n := len(pending)
+	for idx, sid := range pending {
+		if sid == subscriberID {
+			u.pendingSubscriptions[trackSid][idx] = u.pendingSubscriptions[trackSid][n-1]
+			u.pendingSubscriptions[trackSid] = u.pendingSubscriptions[trackSid][:n-1]
+			break
+		}
+	}
+	if len(u.pendingSubscriptions[trackSid]) == 0 {
+		delete(u.pendingSubscriptions, trackSid)
+	}
+}
+
+func (u *UptrackManager) processPendingSubscriptions(resolver func(participantSid string) types.Participant) {
+	updatedPendingSubscriptions := make(map[string][]string)
+	for trackSid, pending := range u.pendingSubscriptions {
+		track := u.getPublishedTrack(trackSid)
+		if track == nil {
+			continue
+		}
+
+		var updatedPending []string
+		for _, sid := range pending {
+			var sub types.Participant
+			if resolver != nil {
+				sub = resolver(sid)
+			}
+			if sub == nil {
```

Comment:
if participant's state is disconnected, we probably also don't want to keep in the list